### ISSUE TITLE
multi region deployment fails on S3 error: Access Denied

### ISSUE THREAD

**Issue by e14nielson** [aws-quickstart/taskcat#315]
ci/taskcat.yml contains:
  package-lambda: true
  regions:
    - us-east-1
    - us-west-2
use taskcat to package lambda code in functions/source/
taskcat will create the s3 bucket in us-east-1
[INFO   ] :Creating bucket taskcat-my-project-temp-f4bc86a3 in us-east-1

when taskcast starts deployment into us-west-2 it fails with S3 error: Access Denied

it looks like the bucket in us-east-1 is not usable by CloudFormation in other regions for deployment.

**Comment by andrew-glenn** [comment]
@e14nielson - Are you seeing this when you use the object URL as the source for the lambda function resource?

*Assuming that's the case*

It's less a taskcat issue and more an S3/Lambda limitation. We've worked around it via a CopyZips lambda... Yes... a lambda that copies lambda zips...

https://github.com/aws-quickstart/quickstart-ribbon-sbc/blob/83dbfd6ecd46d38d1e732cd99a7dbe8d596abce6/templates/ribbon-sbc.template.yaml#L1187-L1291

The "cfnresponse" module is here...

https://github.com/aws-quickstart/quickstart-ribbon-sbc/blob/master/functions/source/KeyGen/cfnresponse.py

Blog post (by @jaymccon) about it here.
- https://aws.amazon.com/blogs/infrastructure-and-automation/deploying-aws-lambda-functions-using-aws-cloudformation-the-portable-way/


*Otherwise...*

Well.. more context please ? :)

**Comment by e14nielson** [comment]
@andrew-glenn  sorry i could have been more clear.

taskcat never gets a template up in us-west-2 as the s3 error is accessing the templates in the deployment bucket not the lambda code. that is just how i have it configured in my project.

it sounds like even once i solve this issue i also have to solve the lambda code issue by deploying a custom lambda to enable taskcat. I wont be doing this which means i probably wont be able to successfully use this feature for my projects.

**Comment by andrew-glenn** [comment]
@e14nielson interesting.  We don't run into this on our end. Sanity check: You're able to manually launch the template via the CFN console using the same params in us-west-2? 

I assume yes, but wanted to make sure. This is puzzling.

**Comment by e14nielson** [comment]
@andrew-glenn everything works fine when ci/taskcat.yml contains:
package-lambda: true
regions:
- us-east-1

as soon as i add us-west-2 it breaks.

**Comment by e14nielson** [comment]
but i can manually deploy to us-west-2, yes

**Comment by avattathil** [comment]
Sounds like you need to add a copyzips lambda to your project to resolve
this issue. The initial build only stages the template in one region (
Which is done by design)

**Comment by e14nielson** [comment]
@avattathil this makes sense but also de-values the multi region feature because of this dependancy. The documentation could be updated to specify this requirement. One of the main reasons i was interested in this was the easy ability to deploy a single stack to all regions without the need for extra infrastructure.

**Comment by avattathil** [comment]
So using taskcat to do the staging (in multiple region s) would make the
template depend on taskcat itself or require end-users to do pre launch
step.

If you implement the staging in the template as described in this blog
https://aws.amazon.com/blogs/infrastructure-and-automation/deploying-aws-lambda-functions-using-aws-cloudformation-the-portable-way/
 It will allow your template to work independently without staging assist
from taskcat.  Once implemented you can use taskcat to test to make sure
everything work as designed in multiple regions.  Hope this helps with your
use case

**Comment by e14nielson** [comment]
again your taskcat docs dont reflect this requirement so that would be great to add.

and there are other ways to do this easier. 
*gitlab runner stages per region

**Comment by andrew-glenn** [comment]
@e14nielson  - QQ: Would an option to auto-generate a bucket per region work for your use-case? In that scenario, there's no cross-region lambda/S3 considerations. I'm happy to add this to the `v0.9.0` release.

**Comment by e14nielson** [comment]
@andrew-glenn that is a great idea, but i also dont understand why a new bucket in each region is needed. a bucket might have regionality in its metadata but the CloudFormation service is able to handle any s3 url provided it has access to.
to test this i put a template in a bucket with region of Oregon then under the CloudFormation service in N. Virginia i can deploy that template as a stack without issue.

**Comment by andrew-glenn** [comment]
@e14nielson It's a lambda specific "gotcha". 

Looking at the [`Code` Object](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-code.html) of  an [`AWS::Lambda::Function` resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-code), we see it'll take 
an S3Bucket, an S3Key, and other properties. The S3Bucket is restricted to the _same region_ as the lambda function. 

> S3Bucket
> An Amazon S3 bucket in the same AWS Region as your function. The bucket can be in a different AWS account.

This restriction is why you'll see CopyZips functions scattered throughout AWS QuickStarts.

BTW: I'm happy to take your feedback on this and forward internally. I just need to get a clear picture of how much pain this causes.

**Comment by andrew-glenn** [comment]
Of course, I just realized the scope of your issue to begin with. (again). What you're specifically running into isn't something that we've experienced, or had others experience either. 

For the sake of usability, I'll add the regional-bucket feature I mentioned above (just to be safe). We're in the back-half of v0.9, so any bugfixes and soforth would show up there. 

I hate to say "hang tight", but we're doing a significant amount of work between v0.8 and v0.9. If this issue persists in v0.9, then I'm more than happy to dig at this deeper.
